src/main/java/com/google/cloud/spanner/pgadapter/utils/MutationWriter.java

```diff
@@ -14,225 +14,580 @@
 
-public class MutationWriter {
+public class MutationWriter implements Callable<Long>, Closeable {
+  public enum CopyTransactionMode {
+    ImplicitAtomic,
+    ImplicitNonAtomic,
+    Explicit,
+  }
 
-  private static final long MUTATION_LIMIT = 20000; // 20k mutation count limit
-  private static final long COMMIT_LIMIT = 100000000; // 100MB mutation API commit size limit
   private static final String ERROR_FILE = "output.txt";
 
-  private boolean hasHeader;
+  private static final int DEFAULT_MUTATION_LIMIT = 20_000; // 20k mutation count limit
+  private static final int DEFAULT_COMMIT_LIMIT =
+      100_000_000; // 100MB mutation API commit size limit
+  /**
+   * The factor that the size of the incoming payload is multiplied with to estimate whether the
+   * current commit size will exceed the maximum commit size. The factor is chosen conservatively to
+   * ensure that a COPY operation of a large file does not unnecessarily break because of an
+   * exceeded commit size.
+   */
+  private static final float DEFAULT_COMMIT_LIMIT_MULTIPLIER_FACTOR = 2.0f;
+
+  private static final int DEFAULT_MAX_PARALLELISM = 128;
+  private static final int DEFAULT_PIPE_BUFFER_SIZE = 1 << 16;
+
+  /**
+   * COPY will INSERT records by default. This is consistent with how COPY on PostgreSQL works. This
+   * option allows PGAdapter to use InsertOrUpdate instead. This can be slightly more efficient for
+   * bulk uploading, and it makes it easier to retry a failed non-atomic batch that might have
+   * already uploaded some but not all data.
+   */
+  private final boolean insertOrUpdate =
+      Boolean.parseBoolean(System.getProperty("copy_in_insert_or_update", "false"));
+
+  private final int commitSizeLimit =
+      Integer.parseInt(
+          System.getProperty("copy_in_commit_limit", String.valueOf(DEFAULT_COMMIT_LIMIT)));
+  private final CopyTransactionMode transactionMode;
+  private final boolean hasHeader;
   private boolean isHeaderParsed;
-  private int mutationCount;
-  private int rowCount;
-  private List<Mutation> mutations;
-  private String tableName;
-  private Map<String, TypeCode> tableColumns;
-  private CSVFormat format;
-  private FileWriter fileWriter;
-  private ByteArrayOutputStream payload = new ByteArrayOutputStream();
+  private long rowCount;
+  private final CloudSpannerJdbcConnection connection;
+  private DatabaseClient databaseClient;
+  private final String tableName;
+  private final Map<String, TypeCode> tableColumns;
+  private final int maxBatchSize;
+  private final long commitSizeLimitForBatching;
+  private final int maxParallelism;
+  private final int pipeBufferSize;
+  private final CSVFormat format;
+  private final CSVParser parser;
+  private PrintWriter errorFileWriter;
+  private final PipedOutputStream payload = new PipedOutputStream();
+  private final AtomicBoolean rollback = new AtomicBoolean(false);
+  private final ListeningExecutorService executorService =
```

Comment:
This executor is responsible for the execution of the separate transactions when using `partitioned_non_atomic`. This significantly speeds up large imports compared to executing all of them sequentially.